File: stream-webrtc-android/src/main/java/io/getstream/webrtc/HardwareVideoEncoderWrapper.java
/*
 * Copyright (c) 2014-2024 Stream.io Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.getstream.webrtc;

/**
 * Original source: https://github.com/shiguredo/sora-android-sdk/blob/3cc88e806ab2f2327bf304207
 * 2e98d6da9df4408/sora-android-sdk/src/main/kotlin/jp/shiguredo/sora/sdk/codec/HardwareVideoEnco
 * derWrapperFactory.kt
 */
class HardwareVideoEncoderWrapper implements VideoEncoder {

    private static final String TAG = "HardwareVideoEncoderWrapper";

    private final VideoEncoder internalEncoder;
    private final int alignment;

    public HardwareVideoEncoderWrapper(VideoEncoder internalEncoder, int alignment) {
        this.internalEncoder = internalEncoder;
        this.alignment = alignment;
    }

    private static class CropSizeCalculator {

        private static final String TAG = "CropSizeCalculator";

        private final int alignment;
        private final int originalWidth;
        private final int originalHeight;
        private final int cropX;
        private final int cropY;

        public CropSizeCalculator(int alignment, int originalWidth, int originalHeight) {
            this.alignment = alignment;
            this.originalWidth = originalWidth;
            this.originalHeight = originalHeight;
            this.cropX = originalWidth % alignment;
            this.cropY = originalHeight % alignment;
            if (originalWidth != 0 && originalHeight != 0) {
                Logging.v(TAG, "init(): alignment=" + alignment +
                        " size=" + originalWidth + "x" + originalHeight + " => " + getCroppedWidth() + "x" + getCroppedHeight());
            }
        }

        public int getCroppedWidth() {
            return originalWidth - cropX;
        }

        public int getCroppedHeight() {
            return originalHeight - cropY;
        }

        public boolean isCropRequired() {
            return cropX != 0 || cropY != 0;
        }

        public boolean hasFrameSizeChanged(int nextWidth, int nextHeight) {
            if (originalWidth == nextWidth && originalHeight == nextHeight) {
                return false;
            } else {
                Logging.v(TAG, "frame size has changed: " +
                        originalWidth + "x" + originalHeight + " => " + nextWidth + "x" + nextHeight);
                return true;
            }
        }
    }

    private CropSizeCalculator calculator = new CropSizeCalculator(1, 0, 0);

    private VideoCodecStatus retryWithoutCropping(int width, int height, Runnable retryFunc) {
        Logging.v(TAG, "retrying without resolution adjustment");
        calculator = new CropSizeCalculator(1, width, height);
        retryFunc.run();
        return VideoCodecStatus.OK;
    }

    @Override
    public VideoCodecStatus initEncode(VideoEncoder.Settings originalSettings, VideoEncoder.Callback callback) {
        calculator = new CropSizeCalculator(alignment, originalSettings.width, originalSettings.height);
        if (!calculator.isCropRequired()) {
            return internalEncoder.initEncode(originalSettings, callback);
        } else {
            VideoEncoder.Settings croppedSettings = new VideoEncoder.Settings(
                    originalSettings.numberOfCores,
                    calculator.getCroppedWidth(),
                    calculator.getCroppedHeight(),
                    originalSettings.startBitrate,
                    originalSettings.maxFramerate,
                    originalSettings.numberOfSimulcastStreams,
                    originalSettings.automaticResizeOn,
                    originalSettings.capabilities
            );
            try {
                VideoCodecStatus result = internalEncoder.initEncode(croppedSettings, callback);
                if (result == VideoCodecStatus.FALLBACK_SOFTWARE) {
                    Logging.e(TAG, "internalEncoder.initEncode() returned FALLBACK_SOFTWARE: " +
                            "croppedSettings " + croppedSettings);
                    return retryWithoutCropping(
                            originalSettings.width,
                            originalSettings.height,
                            () -> internalEncoder.initEncode(originalSettings, callback)
                    );
                } else {
                    return result;
                }
            } catch (Exception e) {
                Logging.e(TAG, "internalEncoder.initEncode() failed", e);
                return retryWithoutCropping(
                        originalSettings.width,
                        originalSettings.height,
                        () -> internalEncoder.initEncode(originalSettings, callback)
                );
            }
        }
    }

    @Override
    public VideoCodecStatus release() {
        return internalEncoder.release();
    }

    @Override
    public VideoCodecStatus encode(VideoFrame frame, VideoEncoder.EncodeInfo encodeInfo) {
        if (calculator.hasFrameSizeChanged(frame.getBuffer().getWidth(), frame.getBuffer().getHeight())) {
            calculator = new CropSizeCalculator(alignment, frame.getBuffer().getWidth(), frame.getBuffer().getHeight());
        }
        if (!calculator.isCropRequired()) {
            return internalEncoder.encode(frame, encodeInfo);
        } else {
            int croppedWidth = calculator.getCroppedWidth();
            int croppedHeight = calculator.getCroppedHeight();
            VideoFrame.Buffer croppedBuffer = frame.getBuffer().cropAndScale(
                    calculator.cropX / 2,
                    calculator.cropY / 2,
                    croppedWidth,
                    croppedHeight,
                    croppedWidth,
                    croppedHeight
            );
            VideoFrame croppedFrame = new VideoFrame(croppedBuffer, frame.getRotation(), frame.getTimestampNs());
            try {
                VideoCodecStatus result = internalEncoder.encode(croppedFrame, encodeInfo);
                if (result == VideoCodecStatus.FALLBACK_SOFTWARE) {
                    Logging.e(TAG, "internalEncoder.encode() returned FALLBACK_SOFTWARE");
                    return retryWithoutCropping(
                            frame.getBuffer().getWidth(),
                            frame.getBuffer().getHeight(),
                            () -> internalEncoder.encode(frame, encodeInfo)
                    );
                } else {
                    return result;
                }
            } catch (Exception e) {
                Logging.e(TAG, "internalEncoder.encode() failed", e);
                return retryWithoutCropping(
                        frame.getBuffer().getWidth(),
                        frame.getBuffer().getHeight(),
                        () -> internalEncoder.encode(frame, encodeInfo)
                );
            } finally {
                croppedBuffer.release();
            }
        }
    }

    @Override
    public VideoCodecStatus setRateAllocation(VideoEncoder.BitrateAllocation allocation, int frameRate) {
        return internalEncoder.setRateAllocation(allocation, frameRate);
    }

    @Override
    public VideoEncoder.ScalingSettings getScalingSettings() {
        return internalEncoder.getScalingSettings();
    }

    @Override
    public String getImplementationName() {
        return internalEncoder.getImplementationName();
    }

    @Override
    public long createNative(long webrtcEnvRef) {
        return internalEncoder.createNative(webrtcEnvRef);
    }

    @Override
    public boolean isHardwareEncoder() {
        return internalEncoder.isHardwareEncoder();
    }

    @Override
    public VideoCodecStatus setRates(VideoEncoder.RateControlParameters rcParameters) {
        return internalEncoder.setRates(rcParameters);
    }

    @Override
    public VideoEncoder.ResolutionBitrateLimits[] getResolutionBitrateLimits() {
        return internalEncoder.getResolutionBitrateLimits();
    }

    @Override
    public VideoEncoder.EncoderInfo getEncoderInfo() {
        return internalEncoder.getEncoderInfo();
    }
}

File: stream-webrtc-android/src/main/java/org/webrtc/HardwareVideoEncoderWrapper.kt
/*
 * Copyright (c) 2014-2023 Stream.io Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.webrtc

/**
 * Original source: https://github.com/shiguredo/sora-android-sdk/blob/3cc88e806ab2f2327bf304207
 * 2e98d6da9df4408/sora-android-sdk/src/main/kotlin/jp/shiguredo/sora/sdk/codec/HardwareVideoEnco
 * derWrapperFactory.kt
 */
internal class HardwareVideoEncoderWrapper(
  private val internalEncoder: VideoEncoder,
  private val alignment: Int,
) : VideoEncoder {
  class CropSizeCalculator(
    alignment: Int,
    private val originalWidth: Int,
    private val originalHeight: Int,
  ) {

    companion object {
      val TAG = CropSizeCalculator::class.simpleName
    }

    val cropX: Int = originalWidth % alignment
    val cropY: Int = originalHeight % alignment

    val croppedWidth: Int
      get() = originalWidth - cropX

    val croppedHeight: Int
      get() = originalHeight - cropY

    val isCropRequired: Boolean
      get() = cropX != 0 || cropY != 0

    init {
      if (originalWidth != 0 && originalHeight != 0) {
        Logging.v(
          TAG,
          "$this init(): alignment=$alignment" +
            "" +
            " size=${originalWidth}x$originalHeight => ${croppedWidth}x$croppedHeight",
        )
      }
    }

    fun hasFrameSizeChanged(nextWidth: Int, nextHeight: Int): Boolean {
      return if (originalWidth == nextWidth && originalHeight == nextHeight) {
        false
      } else {
        Logging.v(
          TAG,
          "frame size has changed: " +
            "${originalWidth}x$originalHeight => ${nextWidth}x$nextHeight",
        )
        true
      }
    }
  }

  companion object {
    val TAG = HardwareVideoEncoderWrapper::class.simpleName
  }

  private var calculator = CropSizeCalculator(1, 0, 0)

  private fun retryWithoutCropping(
    width: Int,
    height: Int,
    retryFunc: () -> VideoCodecStatus,
  ): VideoCodecStatus {
    Logging.v(TAG, "retrying without resolution adjustment")

    calculator = CropSizeCalculator(1, width, height)

    return retryFunc()
  }

  override fun initEncode(
    originalSettings: VideoEncoder.Settings,
    callback: VideoEncoder.Callback?,
  ): VideoCodecStatus {
    calculator = CropSizeCalculator(alignment, originalSettings.width, originalSettings.height)

    if (!calculator.isCropRequired) {
      return internalEncoder.initEncode(originalSettings, callback)
    } else {
      val croppedSettings = VideoEncoder.Settings(
        originalSettings.numberOfCores,
        calculator.croppedWidth,
        calculator.croppedHeight,
        originalSettings.startBitrate,
        originalSettings.maxFramerate,
        originalSettings.numberOfSimulcastStreams,
        originalSettings.automaticResizeOn,
        originalSettings.capabilities,
      )

      try {
        val result = internalEncoder.initEncode(croppedSettings, callback)
        return if (result == VideoCodecStatus.FALLBACK_SOFTWARE) {
          Logging.e(
            TAG,
            "internalEncoder.initEncode() returned FALLBACK_SOFTWARE: " +
              "croppedSettings $croppedSettings",
          )
          retryWithoutCropping(
            originalSettings.width,
            originalSettings.height,
          ) { internalEncoder.initEncode(originalSettings, callback) }
        } else {
          result
        }
      } catch (e: Exception) {
        Logging.e(TAG, "internalEncoder.initEncode() failed", e)
        return retryWithoutCropping(
          originalSettings.width,
          originalSettings.height,
        ) { internalEncoder.initEncode(originalSettings, callback) }
      }
    }
  }

  override fun release(): VideoCodecStatus {
    return internalEncoder.release()
  }

  override fun encode(frame: VideoFrame, encodeInfo: VideoEncoder.EncodeInfo?): VideoCodecStatus {
    if (calculator.hasFrameSizeChanged(frame.buffer.width, frame.buffer.height)) {
      calculator = CropSizeCalculator(alignment, frame.buffer.width, frame.buffer.height)
    }

    if (!calculator.isCropRequired) {
      return internalEncoder.encode(frame, encodeInfo)
    } else {
      // https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/sdk/android/api/org/webrtc/JavaI420Buffer.java;l=172-185;drc=02334e07c5c04c729dd3a8a279bb1fbe24ee8b7c
      val croppedWidth = calculator.croppedWidth
      val croppedHeight = calculator.croppedHeight
      val croppedBuffer = frame.buffer.cropAndScale(
        calculator.cropX / 2,
        calculator.cropY / 2,
        croppedWidth,
        croppedHeight,
        croppedWidth,
        croppedHeight,
      )

      val croppedFrame = VideoFrame(croppedBuffer, frame.rotation, frame.timestampNs)

      try {
        val result = internalEncoder.encode(croppedFrame, encodeInfo)
        return if (result == VideoCodecStatus.FALLBACK_SOFTWARE) {
          Logging.e(TAG, "internalEncoder.encode() returned FALLBACK_SOFTWARE")
          retryWithoutCropping(frame.buffer.width, frame.buffer.height) {
            internalEncoder.encode(
              frame,
              encodeInfo,
            )
          }
        } else {
          result
        }
      } catch (e: Exception) {
        Logging.e(TAG, "internalEncoder.encode() failed", e)
        return retryWithoutCropping(
          frame.buffer.width,
          frame.buffer.height,
        ) { internalEncoder.encode(frame, encodeInfo) }
      } finally {
        croppedBuffer.release()
      }
    }
  }

  override fun setRateAllocation(
    allocation: VideoEncoder.BitrateAllocation?,
    frameRate: Int,
  ): VideoCodecStatus {
    return internalEncoder.setRateAllocation(allocation, frameRate)
  }

  override fun getScalingSettings(): VideoEncoder.ScalingSettings {
    return internalEncoder.scalingSettings
  }

  override fun getImplementationName(): String {
    return internalEncoder.implementationName
  }

  override fun createNativeVideoEncoder(): Long {
    return internalEncoder.createNativeVideoEncoder()
  }

  override fun isHardwareEncoder(): Boolean {
    return internalEncoder.isHardwareEncoder
  }

  override fun setRates(rcParameters: VideoEncoder.RateControlParameters?): VideoCodecStatus {
    return internalEncoder.setRates(rcParameters)
  }

  override fun getResolutionBitrateLimits(): Array<VideoEncoder.ResolutionBitrateLimits> {
    return internalEncoder.resolutionBitrateLimits
  }

  override fun getEncoderInfo(): VideoEncoder.EncoderInfo {
    return internalEncoder.encoderInfo
  }
}

internal class HardwareVideoEncoderWrapperFactory(
  private val factory: HardwareVideoEncoderFactory,
  private val resolutionPixelAlignment: Int,
) : VideoEncoderFactory {
  companion object {
    val TAG = HardwareVideoEncoderWrapperFactory::class.simpleName
  }

  init {
    if (resolutionPixelAlignment == 0) {
      throw java.lang.Exception("resolutionPixelAlignment should not be 0")
    }
  }

  override fun createEncoder(videoCodecInfo: VideoCodecInfo?): VideoEncoder? {
    try {
      val encoder = factory.createEncoder(videoCodecInfo) ?: return null
      return HardwareVideoEncoderWrapper(encoder, resolutionPixelAlignment)
    } catch (e: Exception) {
      Logging.e(TAG, "createEncoder failed", e)
      return null
    }
  }

  override fun getSupportedCodecs(): Array<VideoCodecInfo> {
    return factory.supportedCodecs
  }
}
